File: Stack/Maze/Maze.py
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __eq__(self, other):
        return self.x is other.x and self.y is other.y

    def __add__(self, other):
        return Point(self.x+other.x, self.y+other.y)

    def __repr__(self):
        return '({},{})'.format(self.x, self.y)

    def __iter__(self):
        yield self.x
        yield self.y


class Maze:
    directions = [Point(*xy) for xy in [(-1, 0), (0, 1), (1, 0), (0, -1)]]
    wall = 1
    space = 0

    def __init__(self, maze):
        self.maze = maze

    def __getitem__(self, indices):
        if isinstance(indices, Point):
            x, y = indices
            return self.maze[x][y]
        if isinstance(indices, slice):
            start, end = map(lambda x: Point(
                *x), (indices.start, indices.stop))
            route = [start]

            def go():
                current = route[-1]
                if current == end:
                    return True
                for direction in Maze.directions:
                    point = current + direction
                    if self[point] is Maze.space and point not in route:
                        route.append(point)
                        del point, direction
                        if go():
                            return True
                        route.pop()
                return None
            go()
            return route


if __name__ == "__main__":
    maze = Maze([
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 0, 0, 1, 0, 0, 0, 1, 0, 1],
        [1, 0, 0, 1, 0, 0, 0, 1, 0, 1],
        [1, 0, 0, 0, 0, 1, 1, 0, 0, 1],
        [1, 0, 1, 1, 1, 0, 0, 0, 0, 1],
        [1, 0, 0, 0, 1, 0, 0, 0, 0, 1],
        [1, 0, 1, 0, 0, 0, 1, 0, 0, 1],
        [1, 0, 1, 1, 1, 0, 1, 1, 0, 1],
        [1, 1, 0, 0, 0, 0, 0, 0, 0, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
    ])
    route = maze[(1, 1):(8, 8)]
    print(route)

File: Stack/Maze/Maze.cpp
#include <iostream>
#include <vector>

using namespace std;

const int M = 8, N = 8;

int maze[M + 2][N + 2] = {
    {1, 1, 1, 1, 1, 1, 1, 1, 1, 1},
    {1, 0, 0, 1, 0, 0, 0, 1, 0, 1},
    {1, 0, 0, 1, 0, 0, 0, 1, 0, 1},
    {1, 0, 0, 0, 0, 1, 1, 0, 0, 1},
    {1, 0, 1, 1, 1, 0, 0, 0, 0, 1},
    {1, 0, 0, 0, 1, 0, 0, 0, 0, 1},
    {1, 0, 1, 0, 0, 0, 1, 0, 0, 1},
    {1, 0, 1, 1, 1, 0, 1, 1, 0, 1},
    {1, 1, 0, 0, 0, 0, 0, 0, 0, 1},
    {1, 1, 1, 1, 1, 1, 1, 1, 1, 1},

};

class Point
{
public:
    int x, y;
    Point(int x, int y)
    {
        this->x = x;
        this->y = y;
    }
    Point *add(Point *other)
    {
        Point *point = new Point(0, 0);
        point->x = this->x + other->x;
        point->y = this->y + other->y;
        return point;
    }
    bool eq(Point *other){
        return this->x == other->x && this->y == other->y;
    }
    bool in(vector<Point *> ps){
        for(int i = 0;i < ps.size();i++){
            Point *point = ps[i];
            if(this->eq(point))
                return true;
        }
        return false;
    }
    bool notIn(vector<Point *> ps){
        return !this->in(ps);
    }
    void println(){
        cout << '(' << this->x << ',' << this->y << ')' << endl;
    }
};

Point *start = new Point(1, 1);
Point *stop = new Point(8, 8);
Point *directions[] = {
    new Point(-1, 0),
    new Point(0, 1),
    new Point(1, 0),
    new Point(0, -1),
};
vector<Point *> route;

bool go()
{
    Point *current = route[route.size()-1];
    if(current->eq(stop))
        return true;
    for(int i = 0;i < 4;i++){
        Point *direction = directions[i];
        Point *point = current->add(direction);
        int x = point->x,y = point->y;
        if(maze[x][y] == 0 && point->notIn(route)){
            route.push_back(point);
            if(go())
                return true;
            route.pop_back();
        }
    }
    return false;
}

bool findRoute()
{
    route.clear();
    route.push_back(start);
    return go();
}

int main()
{
    cout << "can find the route of the maze: " << findRoute() << endl;
    for(int i = 0;i < route.size();i++)
        route[i]->println();
}